Add: 힐링 아이템 기능 구현 (HealingItem, HealingItemAttributeSet, GA, GE, Execution)

wjddnjs6520 opened the pull request:
HealingItem
- ConsumableItemBase를 상속받은 힐 아이템 클래스입니다.
- PickupItem 함수를 오버라이드하여 캐릭터의 ASC에 HealingItemAttributeSet을 할당하도록 하였습니다.

HealingItemAttributeSet
- 힐 아이템의 회복량이 들어있는 AttributeSet 입니다.

HealingExecution
- 시전자의 ASC에서 CurrentHealth와 힐아이템의 HealAmount를 불러와 힐 아이템의 회복량만큼 증가시키는 로직을 가지고 있습니다.

GA_UserHealingItem
- Activate Ability 시 캐릭터의 인벤토리의 아이템 개수를 줄이고 GE_HealingItemEffect를 실행합니다.

GE_HealingItemEffect
- HealingExectuion을 할당 받고 로직을 실행합니다.

wjddnjs6520 committed:
Add: 힐링 아이템 기능 구현 (HealingItem, HealingItemAttributeSet, GA, GE, Execution)

HealingItem
- ConsumableItemBase를 상속받은 힐 아이템 클래스입니다.
- PickupItem 함수를 오버라이드하여 HealingItemAttributeSet을 할당하도록 하였습니다.

HealingItemAttributeSet
- 힐 아이템의 회복량이 들어있는 AttributeSet 입니다.

HealingExecution
- 시전자의 ASC에서 CurrentHealth와 힐아이템의 HealAmount를 불러와 힐 아이템의 회복량만큼 증가시키는 로직을 가지고 있습니다.

GA_UserHealingItem
- Activate Ability 시 캐릭터의 인벤토리의 아이템 개수를 줄이고 GE_HealingItemEffect를 실행합니다.

GE_HealingItemEffect
- HealingExectuion을 할당 받아 해당 로직을 실행합니다.

wjddnjs6520 committed:
Merge branch 'Develop' into feature/consumableitem

hwarang98 commented on Source/Shooter/Private/Items/ConsumableItems/HealingExecution.cpp:
```diff
@@ -0,0 +1,31 @@
+	UAbilitySystemComponent* TargetASC = ExecutionParams.GetTargetAbilitySystemComponent();
+	if (!TargetASC) return;
+
+	const UShooterAttributeSet* TargetAttrSet = TargetASC->GetSet<UShooterAttributeSet>();
```
해당 변수는 미사용중인데 나중에 사용하는 로직이 추가될예정일까요?

wjddnjs6520 replied:
DamageExecution에서 복사하면서 생긴 변수인 것 같습니다. 지금은 사용하지 않으니 삭제하겠습니다.

hwarang98 commented on Source/Shooter/Private/Items/ConsumableItems/HealingItem.cpp:
```diff
@@ -0,0 +1,29 @@
+
+	if (HealingItemAttributeSetClass)
+	{
+		UHealingItemAttributeSet* HealingItemAttrSet = NewObject<UHealingItemAttributeSet>(Picker, HealingItemAttributeSetClass);
```
AttributeSet은 보통  ASC 초기화 시점이 1회만 부여하는게 관례입니다! 왜냐하면 다수의 동일 AttributeSet이 추가되면 예기치않은 Attribute충돌, 데이터 중복 발생이 생길수도있기때문입니다! 

```cpp
if (!PickerASC->HasAttributeSet(HealingItemAttributeSetClass))
{
    UHealingItemAttributeSet* HealingItemAttrSet = NewObject<UHealingItemAttributeSet>(Picker, HealingItemAttributeSetClass);
    PickerASC->AddSpawnedAttribute(HealingItemAttrSet);
    UE_LOG(LogTemp, Warning, TEXT("HealingItemAttrSet added to Picker"));
}
```
이런식으로 등록이 되어있지않을때만 UHealingItemAttributeSet를 등록해주는것도 좋을것같습니다!

* 만약 해당로직에 의도된 바 가있으면 그대로 가시죠!

wjddnjs6520 replied:
중복 방지 추가하도록 하겠습니다.

wjddnjs6520 committed:
Refactor: 사용하지 않는 변수 제거, HealingAttributeSet 중복 방지 추가

HealingExecution
- 사용하지 않는 TargetAttrSet 변수 제거

HealingItem
- PickupItem에서 HealingAttributeSet을 중복 방지 로직 추가
- PickerASC에서 HealingItemAttributeSet을 검색하고 해당 AttributeSet이 존재하지 않을 때만 할당하도록 변경
